Document mutation rewrites (#5324)
edgedb/edgedb: docs/datamodel/mutation_rewrites.rst

while updates will set the ``modified`` property:

.. code-block:: edgeql-repl

    db> insert Post {
    ...   title := 'One wierd trick to fix all your spelling errors'
    ... };
    {default::Post {id: 19e024dc-d3b5-11ed-968c-37f5d0159e5f}}
    db> select Post {title, created, modified};
    {
      default::Post {
        title: 'One wierd trick to fix all your spelling errors',
        created: <datetime>'2023-04-05T13:23:49.488335Z',
        modified: {},
      },
    }
    db> update Post
    ... filter .id = <uuid>'19e024dc-d3b5-11ed-968c-37f5d0159e5f'
    ... set {title := 'One weird trick to fix all your spelling errors'};
    {default::Post {id: 19e024dc-d3b5-11ed-968c-37f5d0159e5f}}
    db> select Post {title, created, modified};
    {
      default::Post {
        title: 'One weird trick to fix all your spelling errors',
        created: <datetime>'2023-04-05T13:23:49.488335Z',
        modified: <datetime>'2023-04-05T13:25:04.119641Z',
      },
    }

.. note::

    Each property may have a single ``insert`` and a single ``update`` mutation
    rewrite rule, or they may have a single rule that covers both.

> Special values
> ==============

Reviewer: ```suggestion
Available variables
===================
```

Inside the rewrite rule's expression, you have access to a few special values:

> * ``__subject__`` refers to the new value being set for the property

Reviewer: ```suggestion
* ``__subject__`` refers to the object type with the new property and link values,
```

> * ``__specified__`` is a named tuple with a key for each pointer in the type

Reviewer: ```suggestion
* ``__specified__`` is a named tuple with a key for each property or link in the type
```

  and a boolean value indicating whether this value was explicitly set in the
  mutation

> * ``__old__`` refers to the previous value in update-only mutation rewrite
>   expressions

Reviewer: ```suggestion
* ``__old__`` refers to the object type with the previous property and link values (available for update-only mutation rewrites)
```

Here are some examples of the special values in use. Maybe your blog hosts
articles about particularly controversial topics. You could use ``__subject__``
to enforce a "cooling off" period before publishing a blog post:

.. code-block:: sdl

    type Post {
      ...
      publish_time: datetime {
        rewrite insert, update using (
          __subject__.publish_time ?? datetime_of_statement() +
          cal::to_relative_duration(days := 10)
        )
      }
      ...
    }

Here we take the post's ``publish_time`` if set or the time the statement is
executed and add 10 days to it. That should give our authors time to consider
if they want to make any changes before a post goes live.

You can omit ``__subject__`` in many cases and achieve the same thing:

.. code-block:: sdl-diff

      type Post {
        ...
        publish_time: datetime {
          rewrite insert, update using (
    -       __subject__.publish_time ?? datetime_of_statement() +
    +       .publish_time ?? datetime_of_statement() +
            cal::to_relative_duration(days := 10)
          )
        }
        ...
      }

but only if the path prefix has not changed. In the following schema, for

> example, the ``__subject__`` in the rewrite rule is required since, in the

Reviewer: ```suggestion
example, the ``__subject__`` in the rewrite rule is required, because in the
```

context of the nested ``select`` query, the leading dot resolves from the
``User`` path:

.. code-block:: sdl

    type Post {
      ...
      author_email: str;
      author_name: str {
        rewrite insert, update using (
          (select User {name} filter .email = __subject__.author_email).name
        )
      }
    }
    type User {
      name: str;
      email: str;
    }

.. note::

    Learn more about how this works in our documentation on :ref:`path
    resolution <ref_eql_path_resolution>`.

Using ``__specified__``, we can override ``modified`` with a specified value if
the user provided one in the query while still using our default value
(``datetime_of_statement()``) if they didn't:

.. code-block:: sdl

    type Post {
      ...
      modified: datetime {
        rewrite update using (
          datetime_of_statement()
          if not __specified__.mtime
          else .mtime
        )
      }
      ...

>     }

Reviewer: A more basic example would be updating `modified` only if `title` was changed:

```
      modified: datetime {
        rewrite update using (
          datetime_of_statement()
          if __specified__.title
          else __old__.modified 
        )
      }
```

Just a suggestion, feel free to dismiss.
